Enable CSE on OSSA

meg-gupta opened the pull request.

gottesmm commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -86,6 +88,15 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
 
 public:
+  static SILValue stripCopyValueInst(SILValue val) {
+    auto *inst = val->getDefiningInstruction();
```
A nice little one-liner that does this is:

```
while (auto *cvi = dyn_cast_or_null<CopyValueInst>(val->getDefiningInstruction())) {
  val = cvi->getOperand();
}
```

gottesmm commented:
I started perusing maybe I missed, do you have a bunch of tests with guaranteed values?

meg-gupta commented:
@gottesmm I added some tests with guaranteed values now.

meg-gupta commented:
@swift-ci test

zoecarver commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -86,31 +88,45 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
   using hash_code = llvm::hash_code;
 
 public:
+  static SILValue stripCopyValueInst(SILValue val) {
```
Also, any chance we could try to strip borrows at some point as well?

zoecarver commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -474,22 +566,35 @@ class CSE {
-        RunsOnHighLevelSil(RunsOnHighLevelSil) {}
+        OpenedArchetypesTracker(OpenedArchetypesTracker), Cloner(Cloner),
+        RunsOnHighLevelSil(RunsOnHighLevelSil) {
+    auto &Builder = Cloner.getBuilder();
```
Nit: also no captial. Or maybe even: `Cloner.getBuilder().setOpenedArchetypesTracker...`

meg-gupta commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -859,6 +923,152 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
 
+static void
+collectGuaranteedTransitiveUses(BeginBorrowInst *inst,
+                                SmallVectorImpl<SILInstruction *> &result) {
```
@gottesmm Need your feedback on this. Do you think it'll be useful to move this to BorrowedOperand/Value ?

meg-gupta commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -859,6 +924,148 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
+        if (needsBorrow) {
+          SILBuilderWithScope(end).createEndBorrow(end->getLoc(), borrow);
+        }
+        if (needsCopy && (!lastUser || !hasLifetimeEnd)) {
```
Added some comments to explain

zoecarver replied:
Is there a case where `lastUser == false` and `hasLifetimeEnd == true`? I think this check could be simplified.

zoecarver commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -859,6 +923,152 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
+        // Find if the lastUser of copy was ending its lifetime
+        if (lastUser) {
+          auto lastUserOps = lastUser->getAllOperands();
+          hasLifetimeEnd = std::find_if(
```
Two things: first, I'm surprised this doesn't fail to compile because `std::find_if` returns an iterator. I think this should check `std::find_if(...) != lastUserOps.end()`. Better yet, you could use `std::any_of`. Second, you can use `llvm::find_if`, and then you can pass it a container directly instead of a set of iterators (this applies to all algorithms in this patch).

meg-gupta replied:
yes, thanks! I fixed this instance with std::any_of now.

zoecarver commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -859,6 +923,152 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
+        // Find if the lastUser of copy was ending its lifetime
+        if (lastUser) {
+          auto lastUserOps = lastUser->getAllOperands();
+          hasLifetimeEnd = std::any_of(
```
Nit: you can use `llvm::any_of` and then you don't have to give it both iterators. Sorry for all the back and forth on this piece of code :P

meg-gupta replied:
No! I didn't know about the llvm:: variants. I'll go over everything and fix them.

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci benchmark

meg-gupta commented:
@swift-ci test

meg-gupta committed:
In OSSA rauw utilities use auto generated location for creating copy_value/destroy_value etc

Branch instructions and frontiers can have
LocationKind::ReturnKind/ImplicitReturnKind which are not correct
locations to use for copy_value/destroy_value etc

meg-gupta commented:
@swift-ci test

zoecarver commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -460,10 +507,10 @@ void OwnershipRAUWUtility::eliminateReborrowsOfRecursiveBorrows(
-    auto *innerBorrow =
-        reborrowBuilder.createBeginBorrow(bi->getLoc(), innerCopy);
-    auto *outerEndBorrow = reborrowBuilder.createEndBorrow(bi->getLoc(), value);
+    SILLocation loc = RegularLocation(bi->getLoc().getSourceLoc());
```
Nit: couldn't this just be `bi->getLoc()`?

zoecarver commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -1284,22 +1372,32 @@ class SILCSE : public SILFunctionTransform {
 
-    CSE C(RunsOnHighLevelSil, SEA, FuncBuilder);
+    auto *func = getFunction();
+    SILOpenedArchetypesTracker OpenedArchetypesTracker(func);
```
Nit: `camelCase` not `TitleCase`.

zoecarver reviewed (approved):
This lgtm now. Thanks for all your work on this :)

gottesmm commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -267,21 +301,29 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
+    auto Operands = X->getOperandValues();
+    SmallVector<SILValue, 4> OperandValues;
+    llvm::transform(
+        Operands, std::back_inserter(OperandValues),
```
One thought. Is it possible to create a range that does this transform and pass that to hash_combine instead? That way we do not intermediate arrays (and potential mallocs).

(Keep in mind I am just tagging this one, but I am talking about all of the places that this was done.

meg-gupta replied:
I'll try to do this

meg-gupta replied:
I added a transform function as an optional argument to SILInstruction::getOperandValues. It may be useful in other code as well.

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -73,6 +73,10 @@ Optional<SILBasicBlock::iterator> swift::getInsertAfterPoint(SILValue val) {
   if (isa<SingleValueInstruction>(val)) {
```
why not:

```
if (auto *inst = val->getDefiningInstruction()) {
  return std::next(inst->getIterator());
}
```

gottesmm reviewed (commented):
Some additional comments. I need to read this locally before giving it my +1. I am doing that now.

gottesmm commented:
@meg-gupta I was talking with Andy about this issue around ValueLifetimeAnalysis and returns. I think we are hacking around bugs here. We need a way to have a /real/ location given a terminator insert pt and value lifetime analysis should be smart enough about return inst.

meg-gupta committed:
Add a new overload of SILInstruction::getOperandValues to get
transformed values of the operands

gottesmm commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -201,9 +201,15 @@ CopyValueInst *OwnershipLifetimeExtender::copyAndExtendForNonLifetimeEndingRAUW(
-  bool result = lifetimeAnalysis.computeFrontier(
-      frontier, ValueLifetimeAnalysis::DontModifyCFG, &ctx.deBlocks);
-  assert(result);
+  bool noCriticalEdges = lifetimeAnalysis.computeFrontier(
```
I think this is the wrong way to do this. Why is this needed?

gottesmm replied:
Specifically, if it is needed due to the terminator issue we spoke about before, I think that Andy's recent change adding UnownedInstantaneousUse makes that not necessary for unowned things

meg-gupta replied:
We may have to split the edges while computing frontiers if we have a use on the terminator instruction.

Here's an example:

```
sil [ossa] @struct_extract_test4 : $@convention(thin) (@guaranteed StructWithEnum2) -> () {
bb0(%0 : @guaranteed $StructWithEnum2):
 %1 = struct_extract %0 : $StructWithEnum2, #StructWithEnum2.val
 %copy1 = copy_value %1 : $FakeOptional2
 br bb4

bb4:
 %2 = struct_extract %0 : $StructWithEnum2, #StructWithEnum2.val
 %copy2 = copy_value %2 : $FakeOptional2
 switch_enum %2 : $FakeOptional2, case #FakeOptional2.some1!enumelt:bb5, case #FakeOptional2.some2!enumelt:bb6

bb5(%arg4 : $UInt):
 br bb7(%arg4 : $UInt) --> needs split here

bb6(%arg5 : @guaranteed $Klass):
 %4 = unchecked_trivial_bit_cast %arg5 : $Klass to $UInt
 br bb7(%4 : $UInt)

bb7(%arg6 : $UInt):
 destroy_value %copy2 : $FakeOptional2
 br bb8

bb8:
 destroy_value %copy1 : $FakeOptional2
 %res = tuple ()
 return %res : $()
}
```
Here we are CSE'ing the `struct_extract` at %2 with %1. For this we will have to split the edges at bb5 to get a frontier after the branch.

meg-gupta replied:
You are right, this was not needed. It turned out to be a bug where the ownership rauw was considering transitive uses of none ownership as well.

gottesmm commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -261,8 +261,9 @@ OwnershipLifetimeExtender::copyBorrowAndExtendForRAUW(SILValue newValue,
     SILBuilderWithScope frontierBuilder(insertPt);
-    auto *ebi = frontierBuilder.createEndBorrow(insertPt->getLoc(), borrow);
-    auto *dvi = frontierBuilder.createDestroyValue(insertPt->getLoc(), copy);
+    SILLocation loc = RegularLocation(insertPt->getLoc().getSourceLoc());
```
My suggestion: lets be careful and just use an autogenerated loc. Otherwise sometimes we can get weird stepping behavior sometimes.

gottesmm commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -56,7 +57,34 @@ replaceAllUsesAndEraseInner(SingleValueInstruction *svi, SILValue newValue,
+    // opened archetype is different from copy_value's opened archetype, we have
+    // to clone the copy_value, so that we have the correct opened archetype on
+    // copy_value's result type.
+    bool needsCloning = false;
```
What is causing this to be needed? The reason I am asking is rather than expand what this routine does, I wonder if we could instead put the burden in some way on a different part of the code. Just the context would be useful.

meg-gupta replied:
I removed handling open_existential_ref in CSE. I'll follow this up with another PR with spl handling needed to deal with type dependent operands.

gottesmm commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -527,8 +527,11 @@ class SILInstruction
       OptionalTransformRange<ArrayRef<Operand>, OperandToValue>;
-  OperandValueRange
-  getOperandValues(bool skipTypeDependentOperands = false) const;
+  OperandValueRange getOperandValues(
```
I am hesitant about this part of the change. My gut feeling says that the right thing to do here is to not put this upon all users of this code. Maybe the compiler will optimize it all away... I don't know. Why not instead add a composition type: make a transform range with this property at your local use site.

meg-gupta replied:
I added a new overload of `getOperandValues` without default args, and created a new `TransformedOperandValueRange`.

gottesmm commented on include/swift/SIL/InstructionUtils.h:
```diff
@@ -37,9 +37,13 @@ SILValue stripCasts(SILValue V);
+
+/// Return the underlying SILValue after looking through all copy_value
+/// instructions.
+SILValue lookThroughCopyValueInsts(SILValue v);
```
Can you split out the lookThrough rename into a separate commit? It isn't really part of this commit (I am fine with it in this PR).

gottesmm commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -182,53 +196,70 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
-      llvm::hash_combine_range(Operands.begin(), Operands.end()));
+    auto OperandValues =
+        X->getFunction()->hasOwnership()
+            ? X->getOperandValues(false, lookThroughOwnershipInsts)
```
Why not create a local transform kind here? Why should the needs of CSE impact all use sites of operand values?

gottesmm reviewed (commented):
More initial comments.

meg-gupta committed:
Rename stripOwnershipInsts to lookThroughCopyValueInsts

meg-gupta committed:
Support MultipleValueInstruction in swift::getInsertAfterPoint

meg-gupta committed:
Enable CSE on OSSA

meg-gupta committed:
Do not include transitive uses with none ownership during ownership rauw of guaranteed values

When we are populating transitive users while handling guaranteed values
in ownership rauw, we were including values with none ownership.

Example : rauw of %2 with a dominating value
   Here %arg1 was also considered a transitive use

   %2 = struct_extract %0 : $StructWithEnum2, #StructWithEnum2.val
   %copy = copy_value %2 : $FakeOptional2
   switch_enum %2 : $FakeOptional2, case #FakeOptional2.some1!enumelt:bb5, case #FakeOptional2.some2!enumelt:bb6

  bb5(%arg1 : $UInt):
   br bb7(%arg1 : $UInt)

  bb6(%arg2 : @guaranteed $Klass):
   %4 = unchecked_trivial_bit_cast %arg2 : $Klass to $UInt
   br bb7(%4 : $UInt)

This is incorrect because %arg1 is a trivial value, and this also
leads to ValueLifetimeAnalysis needing a split for finding a frontier
for the use of %arg1 in the branch instruction. In ossa, we should never
have to split edges for finding frontiers, because we do not have critical
edges.

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -70,8 +70,8 @@ const std::function<void(SingleValueInstruction *, SILValue)>
 Optional<SILBasicBlock::iterator> swift::getInsertAfterPoint(SILValue val) {
-  if (isa<SingleValueInstruction>(val)) {
-    return std::next(cast<SingleValueInstruction>(val)->getIterator());
+  if (auto *inst = val->getDefiningInstruction()) {
```
Also, are you using this to insert instructions? Wouldn't SILBuilder::insertAfter work?

gottesmm commented on lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:
```diff
@@ -407,6 +407,9 @@ static void getAllBorrowedValueUsePoints(
     if (auto fOperand = ForwardingOperand::get(usePoints[i])) {
       fOperand->visitForwardedValues([&](SILValue transitiveValue) {
+        // Do not include transitive uses with 'none' ownership
+        if (transitiveValue.getOwnershipKind() == OwnershipKind::None)
```
Nice catch!

gottesmm commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -917,19 +1015,24 @@ bool CSE::processNode(DominanceInfoNode *Node) {
+                                    cast<OpenExistentialRefInst>(AvailInst))) {
+        if (!Inst->getFunction()->hasOwnership() ||
+            Inst->getResults().empty()) {
+          Inst->replaceAllUsesPairwiseWith(AvailInst);
```
Can you move no results to a separate case? Seems weird to me call replaceAllUsesPairwiseWith it since it has no results so no uses!

gottesmm reviewed (commented):
Some more final comments! Looks pretty good! I think we are close!

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci benchmark

meg-gupta commented on lib/SILOptimizer/Transforms/CSE.cpp:
```diff
@@ -267,21 +314,37 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
+          X->getOperandValues(lookThroughOwnershipInsts, false);
+      return llvm::hash_combine(
+          X->getKind(), X->getCallee(),
+          llvm::hash_combine_range(TransformedOperandValues.begin(),
```
I tried this, it doesn't take a range

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -523,12 +523,20 @@ class SILInstruction
   getOperandValues(bool skipTypeDependentOperands = false) const;
+  TransformedOperandValueRange
+  getOperandValues(std::function<SILValue(SILValue)> transformFn,
+                   bool skipTypeDependentOperands) const;
```
Ran into an overload resolution issue with clang when I use a default value for `skipTypeDependentOperands`.
See https://godbolt.org/z/17514r. Leaving it as is for now.

zoecarver replied:
I think the issue is that a pointer picks the implicit bool conversion over the converting constructor in `std::function`. I think it's fine the way it is, but one way to fix it would be to explicitly construct a `std::function` in the caller.

meg-gupta replied:
Yea, it can be a mild surprise. I didn't want to add it to avoid any confusion for new callers.

meg-gupta commented:
@gottesmm I am merging this. I'll add a follow on commit if there are any other comments.

zoecarver commented:
Woohoo! So excited to see ownership making its way through the SILOptimizer. Thanks for all your work on this, @meg-gupta.
